src/ucp/core/ucp_am.c

```diff
@@ -628,10 +705,20 @@ static ucs_status_t ucp_am_bcopy_single_reply(uct_pending_req_t *self)
 
 static ucs_status_t ucp_am_bcopy_multi(uct_pending_req_t *self)
 {
+    ucs_status_t copy_hdr_status;
     ucs_status_t status = ucp_do_am_bcopy_multi(self, UCP_AM_ID_AM_FIRST,
                                                 UCP_AM_ID_AM_MIDDLE,
                                                 ucp_am_bcopy_pack_args_first,
                                                 ucp_am_bcopy_pack_args_mid, 1);
+    ucp_request_t *req   = ucs_container_of(self, ucp_request_t, send.uct);
+
+    if (ucs_unlikely((status == UCS_ERR_NO_RESOURCE) || 
+                     (status == UCP_STATUS_PENDING_SWITCH))) {
+        copy_hdr_status = ucp_am_copy_user_header_to_internal_buf(req);
+        if (ucs_unlikely(copy_hdr_status != UCS_OK)) {
+            return copy_hdr_status;
+        }
+    }
```

Comment:
maybe it could be part of `ucp_am_bcopy_pack_args_first` or `ucp_do_am_bcopy_multi` to simplify the rest

Reply:
Not sure how the pack_cb can help.
In bcopy, UCT calls the pack_cb before the attempt to send, so in the pack_cb we still don't know if the request will go to pending queue.

I can maybe check simpler condition (packed_len == UCS_ERR_NO_RESOURCE) after line 88 in ucp_do_am_bcopy_multi and copy it there.

Reply:
I see now, you are trying to avoid extra copy for user header on fast-path. So it indeed has to be move to `ucp_do_am_bcopy_multi` under `offset == 0` condition

Reply:
Hi. I look again on ucp_do_am_bcopy_multi.
This function is used by other protocols that rely on AM API, see my previous [discussion](https://github.com/openucx/ucx/pull/8173#discussion_r875614275) with Mikhail.

I'm not sure if adding branches that are specific for AM protocol is a good idea

Reply:
inlining of `ucp_do_am_bcopy_multi` assumes that all compile time branches must be optimized in compile time